### PR TITLE
Added support for additional disks; Windows Server 2008R2 support; Ad…

### PR DESCRIPTION
**Pull request by stiller-leser**
…ded support for scratch disks

@jjasghar Couldn't create full commit history anymore since I deleted my fork. Here are the last commits that are relevant. This also solves #19 for my use cases as well. Happy to change the versioning, turn into one version if desired.

On the plus side it is cleaner than the original #61 and resolves some bugs we have found in the meantime (disks not getting deleted etc).

### CODE REVIEW

**Comment by jjasghar** [comment]
@robbkidd any thoughts on this?

**Comment by stiller-leser** [comment]
Is there any progress here? Happy to include changes if needed.

**Comment by stiller-leser** [comment]
I reverted the changelog and versioning. It's now failing because of the mentioned bundler issue.

**Comment by stiller-leser** [comment]
What's the satus here? Anything I can provide / do from my side?

File: lib/kitchen/driver/gce.rb [hunk 1]
```diff
@@ -141,26 +141,97 @@ def destroy(state)
...
+          autodelete_disk: true,
+          disk_size: 10,
+          disk_type: "pd-standard",
+        }
+
+        if old_disk_configuration_present?
+          # If the old disk configuration is used,
+          # we'll convert it to the new one
+          config[:disks] = {}
+          config[:disks][server_name] = {
```
Comment (robbkidd) [inline lib/kitchen/driver/gce.rb]: The use of `server_name` here results in an error `undefined local variable or method 'server_name'`. This should probably be `:disk1` instead.

It's a bit confusing, but `server_name` in this class is not an instance method. Everywhere it is used, `server_name` is a local variable for that method. The value is generated once and stored in `state[:server_name]` and from there it is retrieved and set again for other methods that need to know the name.

I don't think this `create_disks_config` method needs to use the server name, though. Using `:disk1` here will work, much like the final case in this method where there is no explicit disk configuration, so a disk config is created named `disk1`.

File: lib/kitchen/driver/gce.rb [hunk 2]
```diff
@@ -141,26 +141,97 @@ def destroy(state)
...
+            boot: true,
+          }
+          config[:disks][server_name][:autodelete_disk] = config[:autodelete_disk].nil? ? disk_default_config[:autodelete_disk] : config[:autodelete_disk]
+          config[:disks][server_name][:disk_size] = config[:disk_size].nil? ? disk_default_config[:disk_size] : config[:disk_size]
+          config[:disks][server_name][:disk_type] = config[:disk_type].nil? ? disk_default_config[:disk_type] : config[:disk_type]
```
Comment (robbkidd) [inline lib/kitchen/driver/gce.rb]: Style recommendation:

The creation of a new_disk_configuration based on an old_disk_configuration can be simplified and made maybe a little easier to read with the [`Hash#fetch`](https://ruby-doc.org/core-2.5.0/Hash.html#method-i-fetch) method.

```ruby
# If the old disk configuration is used,
# we'll convert it to the new one
config[:disks] = {
  disk1: {
    boot: true,
    autodelete_disk: config.fetch(:autodelete_disk, disk_default_config[:autodelete_disk]),
    disk_size: config.fetch(:disk_size, disk_default_config[:disk_size]),
    disk_type: config.fetch(:disk_type, disk_default_config[:disk_type]),
  }
}
```

File: lib/kitchen/driver/gce.rb [hunk 3]
```diff
@@ -141,26 +141,97 @@ def destroy(state)
...
+          raise "Disk type #{config[:disks][server_name][:disk_type]} is not valid" unless valid_disk_type?(config[:disks][server_name][:disk_type])
+        elsif new_disk_configuration_present?
+          # If the new disk configuration is present, ensure that for
+          # every disk the needed configuration is set
+          boot_disk_counter = 0
+          config[:disks].each do |disk_name, disk_config|
+            # te&/ => te
+            raise "Disk name invalid. Must match #{DISK_NAME_REGEX}." if disk_name.to_s.match(DISK_NAME_REGEX).to_s.length < disk_name.length
```
Comment (robbkidd) [inline lib/kitchen/driver/gce.rb]: Recommend moving the `if` logic here and the `DISK_NAME_REGEX` to a `valid_disk_name?` method, in the same way the logic for disk types is wrapped in `valid_disk_type?`

File: README.md [hunk 4]
```diff
@@ -300,6 +280,62 @@ Default:
     "test-kitchen-instance" => <instance.name>,
     "test-kitchen-user"     => <env_user>,
 
+### Disk configuration
```
Comment (robbkidd) [inline README.md]: I squeed a little that there were doc updates included. 💞 

squeeing

File: lib/kitchen/driver/gce.rb (hunk 1 above)
Comment (robbkidd) [reply lib/kitchen/driver/gce.rb]: The current situation has `server_name` passed in to the `boot_disk()` method from [`create_instance_object`](https://github.com/test-kitchen/kitchen-google/blob/master/lib/kitchen/driver/gce.rb#L330) which itself had `server_name` passed in from the [`create`](https://github.com/test-kitchen/kitchen-google/blob/master/lib/kitchen/driver/gce.rb#L101) method where the name was [generated and stored as a local variable](https://github.com/test-kitchen/kitchen-google/blob/master/lib/kitchen/driver/gce.rb#L98). It's not _clean_, by any means.

The server name is not necessary here where the disk configuration is being built. It only needs a sensible short name for the disk, similar to the ones written in the README that we expect a human to write into their kitchen.yml. With your new `create_disks` method, [the server name and the disk name are concatenated](https://github.com/test-kitchen/kitchen-google/pull/62/files#diff-6cfd6bc13548320b2c96b01086644ddcR431) for the name used for the "real" disk created in GCE. So giving this disk a short, sensible default name like "disk1" will work. (It did when I tested [the simplified transform from old to new in one of my other comments](https://github.com/test-kitchen/kitchen-google/pull/62#discussion_r183055011).)

**Pull request merged**